rust/common/src/value.rs

```diff
@@ -157,17 +157,62 @@ impl_tvm_context!(
     DLDeviceType_kDLExtDev: [ext_dev]
 );
 
+/// A struct holding TVM byte-array.
+///
+/// ## Example
+///
+/// ```
+/// let v = b"hello";
+/// let barr = TVMByteArray::from(&v);
+/// assert_eq!(barr.len(), v.len());
+/// assert_eq!(barr.data(), 7[104u8, 101, 108, 108, 111]);
+/// ```
 impl TVMByteArray {
```

Comment:
`TVMByteArray` needs a lifetime if it's storing unowned data, but `TVMByteArray` is semantically a `&[u8]` and shouldn't be exposed to clients. Its only use is for FFI, so I'd suggest adding converters from `&[u8]`

Reply:
Frontend needs to pass the loaded params to the registered `load_param_fn`

```
let params: Vec<u8> = fs::read("deploy_param.params").unwrap();
let barr = TVMByteArray::from(&params);
call_packed!(load_param_fn, &barr).unwrap();
```

 The `data` method is outputing `'static [u8]` and we have `impl<T: AsRef<[u8]>> From<T> for TVMByteArray` defined, so you mean it's better to make two conversions instead, like

```
impl<'a> From<&'a [u8]> for TVMByteArray
```
and 

```
impl From<Vec<u8>> for TVMByteArray
```
But the [`as_ref`](https://doc.rust-lang.org/std/convert/trait.AsRef.html#tymethod.as_ref) does it under the hood.

Reply:
sure, whatever `From`s are necessary. The point is that `TVMByteArray` is just an unsafe `&[u8]` that's used solely for FFI. Expositing it to the client is decidedly ungood.

Also, the semantics of the following snippet don't make sense.

```rust
let params: Vec<u8> = fs::read("deploy_param.params").unwrap();
let barr = TVMByteArray::from(&params);
```

`TVMByteArray` is an array of bytes but, in the general, a `.params` file contains many arrays.
I think that you want something like `runtime::load_params(...) -> Vec<Tensor>`. What's the story for `Tensor` in frontend? We should consider moving that to `common`, too.

Reply:
That's mimicking from C++ example. `.params` is a byetarray and similar conversion exists already. NDArray cannot be put in common because they're only for frontend. Also I was thinking for the sake of compatibility, it's better for runtime to rename Tensor to NDArray overall.

Reply:
> `.params` is a byetarray

Huh? How are you getting this `.params` file? It's certainly not produced by [`tvm.relay.save_param_dict`](https://github.com/dmlc/tvm/blob/cffb4fba03ea582417e2630bd163bca773756af6/src/relay/backend/param_dict.cc#L39-L70)...

Reply:
Not about the relay, but it's been generated from [`nnvm.compiler.save_param_dict`](https://github.com/dmlc/tvm/blob/master/rust/frontend/examples/resnet/src/build_resnet.py#L80) and here's the C++ [example](https://docs.tvm.ai/deploy/nnvm.html).

Reply:
> Not about the relay

You'll notice that [the implementation is actually the same](https://github.com/dmlc/tvm/blob/cffb4fba03ea582417e2630bd163bca773756af6/nnvm/src/compiler/graph_runtime.cc#L57-L87). I think that you're confusing `TVMByteArray` with [Python `bytearray`](https://docs.python.org/3/library/functions.html?highlight=bytearray#func-bytearray) which simply converts the loaded file into a mutable `bytes` object.

>  not a good name in general

A fair point. Combatting industry hype is a worthy objective!